masterplan_tools/method/blocks/blocks_clustering.py

```diff
@@ -0,0 +1,415 @@
+        inner_poly = gpd.GeoDataFrame(geometry=[rectangle], crs=self.local_crs)
+
+        return inner_poly
+
+    def get_inner_geom(self, my_inner_poly, new_inner_polygons_list, i, inner_polys, if_break=False):
+        i = i.geometry.item()
+
+        if my_inner_poly.representative_point().item().intersects(i):
+            inner_poly_boundary2 = self.get_connection_lines(my_inner_poly, i, inner_polys)
+
+            inner_poly_boundary2 = inner_poly_boundary2.convex_hull
+            res = i.difference(inner_poly_boundary2)
+            new_inner_polygons_list.append(
+                gpd.GeoDataFrame(geometry=[i.intersection(inner_poly_boundary2)], crs=self.local_crs)
+            )
+
+            res = gpd.GeoDataFrame(geometry=[res], crs=self.local_crs)
+
+            if isinstance(res["geometry"].item(), MultiPolygon):
+                res = res.explode().reset_index(drop=True)
+
+            if res.shape[0] > 1:
+                for i in res.index:
+                    new_inner_polygons_list.append(gpd.GeoDataFrame(geometry=res.loc[i, :], crs=self.local_crs))
+            else:
+                new_inner_polygons_list.append(res)
+            if_break = True  # FIXME is it right?
+
+            return if_break
+        return False
+
+    def clusterize_blocks(self):
+        old_new_blocks = []
+        new_polys = []
+        bigger_polytmp = []
+
+        for c, block_id in enumerate(tqdm(self.blocks_to_consider)):
+            inner_polys = []
+            new_inner_polygons_list = []
+
+            t_build = self.buildings_centroids[self.buildings_centroids["id"] == block_id].copy()
+            t_build["Latitude"] = t_build.geometry.x
+            t_build["Longitude"] = t_build.geometry.y
+
+            houses_cluster = self.get_clusters(t_build)
+            bigger_poly = self.blocks[self.blocks["id"] == block_id]
+            houses_cluster = houses_cluster[houses_cluster["cluster"] != -1]
+
+            if len(set(houses_cluster["cluster"])) == 1 and houses_cluster["cluster"].shape[0] > 1:
+                new_polys.append(houses_cluster)
+                continue
+
+            for _ in set(houses_cluster["cluster"]):
+                # Create spatial index
+                gdf_sindex = houses_cluster.sindex
+                line = bigger_poly.boundary.item()
+
+                # Get nearest geometry
+                nearest_geom_index = list(gdf_sindex.nearest(line, 1))[0]
+                cluster = houses_cluster.iloc[nearest_geom_index]["cluster"].item()
+
+                try:
+                    inner_polygon = self.get_inner_poly(houses_cluster, cluster)
+                    inner_polys.append(inner_polygon)
+                except AttributeError:
+                    print("AttributeError happened")
+                    continue
+
+                houses_cluster = houses_cluster[houses_cluster["cluster"] != cluster]
+
+            for inner_polygon in inner_polys:
+                flag = 0
```

Comment:
The `bool` type for this flag, and why not to give it a convenient name to understand what it indicates?